text/0002-extensible-provisioning-templates.md

```diff
@@ -0,0 +1,118 @@
+-   Feature Name: Extensible provisioning templates
+-   Start Date: 2016-06-28
+-   RFC PR: (leave this empty)
+-   Issue: (leave this empty)
+
+# Summary
+
+[summary]: #summary
+
+This is a suggestion for refactoring Foreman's template generation mechanism is such a way that will enable
+different plugins to add plugin-specific code to the rendered template.
+
+# Motivation
+
+[motivation]: #motivation
+
+## Original use case
+
+As part of the effort to clean up Foreman's core from puppet related code,
+I have stumbled upon `Host::Managed#handle_ca` method. This method is responsible
+for dealing with certificate autosigning mechanism. The method is called from different parts
+in the code: from `UnattendedController` and from two orchestration concerns: `Orchestration::SSHProvision`
+and `Orchestration::Compute`. The motivation for this method is to keep the autosigning window open
+for as shortest period of time as possible.
+My suggestion is to solve this design flaw by using extensible templates in the following process:
+
+1.  Foreman will generate a certificate during the regular orchestration process and store it as part of
+    host's information
+2.  Finish template will be extended by the puppet plugin so it will install the certificate to the newly
+    created host.
+
+This method has many advantages that are out of scope for this RFC.
+
+## Advantages of extensible templates
+
+1.  Encapsulation of plugin related template code inside the plugin: no more templates that are not compatible
+    with certain versions of a plugin.
+2.  Providing a valid extension point for actions that should be executed upon actual host build.
+
+# Detailed design
+
+[design]: #detailed-design
+
+We will use existing mechanism for extending views called [pagelets](http://projects.theforeman.org/projects/foreman/wiki/Pagelets).
+We will add a call to render pagelets that are already registered each time a template is rendered and pass all the information about the template.
+Any plugin will be able to register its own partial that will be embedded into the final template.
+
+## Changes to the core
+
+1.  Add a call to pagelets to the template sent for rendering in `UnattendedController#safe_render`:
+
+```ruby
+render :inline => "<%= unattended_render(@unsafe_template, @template_name).html_safe %> <%= render_pagelets_for(:template_extension, :subject => @template_object) %>" and return
```

Comment:
>we need to tell the plugins what template is being rendered

That's the reason you have `:subject => @template_object` - it gives the plugin access to `Template` object, so it can examine it. You can see the example in "Usages" section, I am doing exactly this kind of filtering by using `:onlyif` option provided by the pagelet.